src/openms/source/FORMAT/DATAACCESS/S3ChunkedInputSource.cpp

```diff
@@ -0,0 +1,483 @@
+                            unsigned long bytesRead = remainingBytes - m_zStream.avail_out;
+                            totalBytesRead += bytesRead;
+                            m_position += bytesRead;
+                            remainingBytes = m_zStream.avail_out;
+                            break;
+                        }
+                        else if (ret == Z_OK) {
+                            // toFill is full
+                            unsigned long bytesRead = remainingBytes - m_zStream.avail_out;
+                            totalBytesRead += bytesRead;
+                            m_position += bytesRead;
+                            remainingBytes = m_zStream.avail_out;
+                            break;
+                        }
+                        else if (ret != Z_BUF_ERROR) {
+                            OPENMS_LOG_ERROR << "Error occurred during decompression. Zlib error code: " << ret << std::endl;
+                            break;
+                        }
+                    } while (true);
+                    return static_cast<XMLSize_t>(totalBytesRead);
+                }
+                // we reached the end of the current chunk
+                // load next chunk
+                m_req.SetRange("bytes=" + std::to_string(m_currentChunkEnd + 1) + "-" + std::to_string(m_currentChunkEnd + m_chunkSize));
+                m_currentChunkEnd = m_currentChunkEnd + m_chunkSize;
+                Aws::S3::Model::GetObjectOutcome outcome = m_client.GetObject(m_req);
+                if (outcome.IsSuccess()) {
+                    m_currentChunk = outcome.GetResultWithOwnership();
+                    currentBodyStream = &m_currentChunk.GetBody();
+                } else {
+                    OPENMS_LOG_ERROR << "Error loading chunk from S3: " << outcome.GetError().GetExceptionName() << ": " << outcome.GetError().GetMessage() << std::endl;
+                }
+            }
+            else
+            { // neither good nor eof -> error
+                OPENMS_LOG_ERROR << "Error in AWS S3 SDK result stream: " << currentBodyStream->rdstate() << std::endl;
+            }
+        }
+        return static_cast<XMLSize_t>(totalBytesRead);
+    }
+
+    const XMLCh* S3ChunkedGzipBinInputStream::getContentType() const {return nullptr;};
+
+    S3ChunkedBzip2BinInputStream::S3ChunkedBzip2BinInputStream(const Aws::S3::S3Client& client, const Aws::S3::Model::GetObjectRequest& req, unsigned long chunkSize)
+        : m_req(req), m_client(client), m_position(0), m_chunkSize(chunkSize), m_decompressedBuffer(), m_bzStream()
+    {
+        // Create a GZIP decompression stream
+        m_bzStream.bzalloc = NULL;
+        m_bzStream.bzfree = NULL;
+        m_bzStream.opaque = NULL;
+        m_bzStream.avail_in = 0;
+        m_bzStream.next_in = NULL;
+
+        if (BZ2_bzDecompressInit(&m_bzStream, 0, 0) != BZ_OK) {
+            throw new std::runtime_error("Error occurred during bzip decompression");
```

Comment:
now need to dynamically allocate a new object. Just throw the thing!

(more occurrences below)